### DISCUSSION TITLE
Questions about BDDL construction

### DISCUSSION THREAD

**Discussion by Daniel70-liu** [StanfordVL/BEHAVIOR-1K#1006] [Q&A]
Hello, I referred to the question and answer of #972, but I still have some things that I am not clear about.
1. What is agent.n.01 that appears in every bddl file?    (ontop agent.n.01_1 floor.n.01_1)  Does it refer to the robot?
2. Is it not allowed to include parameters like "afhwym" from "grocery_shelf-afhwym" in the bddl?
3. Can you guide me on how to make the robot in Omnigibson take cookies and wine from the shelf and place them into a shopping basket at the checkout counter?
The following bddl file is the functionality I want to implement.
```
(define (problem demo_pick_task-0)
    (:domain omnigibson)

    (:objects
	shelf.n.01_1 shelf.n.01_2 - shelf.n.01
	checkout.n.03_1 - checkout.n.03
	shopping_basket.n.01_1 - shopping_basket.n.01
        box__of__cookies.n.01_1 - box__of__cookies.n.01
        bottle__of__wine.n.01_1 - bottle__of__wine.n.01
	floor.n.01_1 - floor.n.01
        agent.n.01_1 - agent.n.01
    )
    
    (:init 
	(ontop shopping_basket.n.01_1 checkout.n.03_1)
        (inside box__of__cookies.n.01_1 shelf.n.01_1)
        (inside bottle__of__wine.n.01_1 shelf.n.01_2) 
	(inroom floor.n.01_1 grocery_store) 
        (inroom shelf.n.01_1 grocery_store) 
        (inroom shelf.n.01_2 grocery_store) 
        (inroom checkout.n.03_1 grocery_store)
	(ontop agent.n.01_1 floor.n.01_1)
    )
    
    (:goal 
        (and 
            (inside ?box__of__cookies.n.01_1 ?shopping_basket.n.01_1)
	    (inside ?bottle__of__wine.n.01_1 ?shopping_basket.n.01_1)
	    (ontop ?shopping_basket.n.01_1 ?checkout.n.03_1)
        )
    )
)
```
Thank you very much for reading my question.

**Comment by hang-yin** [comment]
Hi @Daniel70-liu !

Again, I'm not a BDDL expert, but I can give you my limited understanding. @ChengshuLi please correct me if I make any mistakes. 

1. You are absolutely correct, `agent.n.01` refers to the robot. `(ontop agent.n.01_1 floor.n.01_1)` essentially means we want our robot to be on top of `floor.n.01_1`. By specifying where `floor.n.01_1` belongs to, e.g. living room, we would thus constrain the robot to be in this room as well. 
2. Parameters like `afhwym` is a model identifier of one specific object. However, in the case of BDDL, we are specifying abstract relationships between object instances, thus we do not need/want to specify the specific model identifier. For example, in the case of domain randomization, we need each object instance specified in the BDDL to be swappable with other object models. If you want a very specific task initial set up, you can sample it with online sampling multiple times, grab the one you want, and save it for future use. 
3. Do you mean writing the bddl initial condition for the task you are describing or actually completing this task? Your bddl looks generally correct, are you asking because you are having trouble sampling this? If you are looking for ways to complete this task with physical actions from the robot, we do not yet have robot trajectories for these. While we are actively working on these data, you can try to chain our action primitives (implemented with motion planning) together to finish a specific task. You will need to use docker installation for OMPL though.

**Reply by Daniel70-liu** [reply]
@hang-yin Thank you for your help, the BDDL file has successfully passed custom task validation. However, I encountered some issues during sampling. The program below modifies the behavior task example with **"scene_model": "grocery_store"** and **"activity_name": "demo_pick_task"**, but I ran into an error displayed in the image below. How should I resolve it?

```
import omnigibson as og
cfg = {
    "scene": {
        "type": "InteractiveTraversableScene",
        "scene_model": "grocery_store",
    },
    "robots": [
        {
            "type": "Fetch",
            "obs_modalities": ["rgb"],
            "default_arm_pose": "diagonal30",
            "default_reset_mode": "tuck",
        },
    ],
    "task": {
        "type": "BehaviorTask",
        "activity_name": "demo_pick_task",
        "activity_definition_id": 0,
        "activity_instance_id": 0,
        "online_object_sampling": True,
    },
}
env = og.Environment(configs=cfg)
env.task.save_task()
```

Error log:
[ERROR] [omnigibson.utils.bddl_utils] initial: Room type [grocery_store] of scene [grocery_store_half_stocked] do not contain or cannot sample all the objects needed.
The following are the possible room instances for each object, the intersection of which is an empty set.
floor.n.01_1: grocery_store_0
shelf.n.01_1: grocery_store_0
shelf.n.01_2: grocery_store_0
checkout.n.03_1:

2024-10-26 12:58:13 [1,790,216ms] [Error] [omnigibson.utils.bddl_utils] initial: Room type [grocery_store] of scene [grocery_store_half_stocked] do not contain or cannot sample all the objects needed.
The following are the possible room instances for each object, the intersection of which is an empty set.
floor.n.01_1: grocery_store_0
shelf.n.01_1: grocery_store_0
shelf.n.01_2: grocery_store_0
checkout.n.03_1:
![{A02F427D-0787-431E-9B06-9B3502558DE6}](https://github.com/user-attachments/assets/5d29ba76-6dc0-44c8-ab0a-20e293985536)

**Reply by hang-yin** [reply]
Hi @Daniel70-liu , I think what's happening is that your object category names are not matching the actual objects in the scene. Here's the link to some details about this scene: https://behavior.stanford.edu/knowledgebase/scenes/grocery_store_half_stocked/index.html

It seems to me that `shelf` should be something like `grocery_shelf` and `checkout` might need to be something like `checkout_counter`. These need to match the objects in the scene exactly as they show up in the knowledgebase because these are specified with the `inroom` condition.

**Reply by Daniel70-liu** [reply]
Hello @hang-yin , After solving the scene model configuration issue, I can now successfully sample and output json files. I changed "scene_model": "grocery_store" to "scene_model": "grocery_store_half_stocked". But the error still appears and I'm not sure if it actually worked.

**Reply by hang-yin** [reply]
@Daniel70-liu do you mind posting your new error log here? I can also help look into this more if you can post your updated BDDL file.

**Reply by Daniel70-liu** [reply]
@hang-yin Sure, here are my log and BDDL files.I have one more question about the robot camera. In the YAML configuration file, I specified that the robot should use **obs_modalities: [scan, bbox_3d]**. However, after running it, there was no change. Could there be a setting I missed?

Log:
[log.txt](https://github.com/user-attachments/files/17627193/log.txt)
BDDL files:
[grocery_store_half_stocked_task_demo_pick_task_0_0_template.json](https://github.com/user-attachments/files/17627158/grocery_store_half_stocked_task_demo_pick_task_0_0_template.json)

Fetch_robot.yaml:
![{AC208BF1-27EA-4222-82AA-3A63D0A974C5}](https://github.com/user-attachments/assets/cdf520c3-323b-4b08-bec8-110519389a42)

**Reply by Daniel70-liu** [reply]
Hi @hang-yin, sorry to bother you, but could you help me out? Thanks so much!
